toolchain/sem_ir/ids.h

```diff
@@ -957,6 +957,25 @@ struct LocId : public IdBase, public Printable<LocId> {
 
 constexpr LocId LocId::Invalid = LocId(Parse::NodeId::Invalid);
 
+// Polymorphic id for fields in `Any[...]` typed instruction groups. Used for
+// fields where the specific instruction structs have different field types in
+// that position or do not have a field in that position at all. Allows
+// conversion with `Inst::As<>` from the specific typed instruction to the
+// `Any[...]` instruction group.
+//
+// This type participates in `Inst::FromRaw` in order to convert from specific
+// instructions to an `Any[...]` instruction group, but does not participate in
+// `Inst::ToRaw` as it's not possible to convert in the other direction:
+// - In the case the specific instruction has a field of some `IdKind` in the
+//   same position, the `Any[...]` type will hold its raw value in the
+//   `AnyRawId` field.
+// - In the case the specific instruction has no field in the same position, the
+//   `Any[...]` type will hold a default constructed `AnyRawId` with an invalid
+//   value.
+struct AnyRawId {
+  int32_t raw_id = IdBase::InvalidIndex;
+};
```

Comment:
Can we make this derive from `IdBase`, so that we don't need custom `ToRaw` / `FromRaw`? I wonder if we can also make `IntId` do the same, so there are no exceptions to the rule that field types derive from `IdBase`...

If there's some reason that's undesirable / doesn't work, I think this is fine as-is.

Reply:
Yeah I think that makes sense. If `IdBase` contained a bunch of id-based API then I would feel less good about it, as this represents a kind of type-erased Id so really shouldn't be used as an Id itself, but it just moves the storage to the base class, provides Printable, and requires adding two constructors instead in `AnyRawId`.